core/types/src/key_request.rs

```diff
@@ -0,0 +1,273 @@
+    pub const ISV_FAMILY_ID: KeyPolicy = KeyPolicy(SGX_KEYPOLICY_ISVFAMILYID);
+
+    /// Derive key with the enclave's ISVEXTPRODID
+    pub const ISV_EXTENDED_PROD_ID: KeyPolicy = KeyPolicy(SGX_KEYPOLICY_ISVEXTPRODID);
+}
+
+impl core::ops::BitOr<KeyPolicy> for KeyPolicy {
+    type Output = Self;
+    #[inline]
+    fn bitor(self, other: Self) -> Self {
+        KeyPolicy(self.0 | other.0)
+    }
+}
+impl ::core::ops::BitOrAssign for KeyPolicy {
+    #[inline]
+    fn bitor_assign(&mut self, rhs: KeyPolicy) {
+        self.0 |= rhs.0;
+    }
+}
+impl ::core::ops::BitAnd<KeyPolicy> for KeyPolicy {
+    type Output = Self;
+    #[inline]
+    fn bitand(self, other: Self) -> Self {
+        KeyPolicy(self.0 & other.0)
+    }
+}
+impl ::core::ops::BitAndAssign for KeyPolicy {
+    #[inline]
+    fn bitand_assign(&mut self, rhs: KeyPolicy) {
+        self.0 &= rhs.0;
+    }
+}
+
+new_type_wrapper! {
+    ConfigSvn, sgx_config_svn_t;
+}
+
+// Suppress clippy as the `new_type_wrapper` macro can't derive Default for many
+// of the types.
+#[allow(clippy::derivable_impls)]
+impl Default for ConfigSvn {
+    fn default() -> Self {
+        ConfigSvn(0)
+    }
+}
+
+new_type_wrapper! {
+    IsvSvn, sgx_isv_svn_t;
+}
+
+#[allow(clippy::derivable_impls)]
+impl Default for IsvSvn {
+    fn default() -> Self {
+        IsvSvn(0)
+    }
+}
+
+new_type_wrapper! {
+    CpuSvn, sgx_cpu_svn_t;
+}
+
+impl Default for CpuSvn {
+    fn default() -> Self {
+        CpuSvn(sgx_cpu_svn_t {
+            svn: [0; SGX_CPUSVN_SIZE],
+        })
+    }
+}
+
+new_type_wrapper! {
+    KeyRequest, sgx_key_request_t;
+}
+
+impl KeyRequest {
+    /// Creates a new [KeyRequest].
+    ///
+    /// The [KeyRequest] will be mostly empty and can be built up using the
+    /// setter methods like [KeyRequest::set_isv_svn]
+    ///
+    /// # Arguments
+    ///
+    /// * `csprng` - A cryptographic psuedo random number generator used to
+    ///   ensure similar key requests don't collide.
+    pub fn new<R: CryptoRng + RngCore>(csprng: &mut R) -> Self {
+        let mut wear_out_protection = [0u8; SGX_KEYID_SIZE];
+        csprng.fill_bytes(&mut wear_out_protection);
+
+        KeyRequest(sgx_key_request_t {
+            key_name: 0,
+            key_policy: 0,
+            isv_svn: 0,
+            reserved1: 0,
+            cpu_svn: CpuSvn::default().into(),
+            attribute_mask: Attributes::default().into(),
+            key_id: sgx_key_id_t {
+                id: wear_out_protection,
+            },
+            misc_mask: MiscellaneousSelect::default().into(),
+            config_svn: 0,
+            reserved2: [0; SGX_KEY_REQUEST_RESERVED2_BYTES],
+        })
+    }
+}
+
+impl KeyRequest {
+    /// Set the key name to use in the key request
+    ///
+    /// # Arguments
+    ///
+    /// * `key_name` - The key name to use
+    pub fn set_key_name(mut self, key_name: KeyName) -> Self {
+        self.0.key_name = ::core::intrinsics::discriminant_value(&key_name);
```

Comment:
If all you want is to turn key_name into a u16, why not just use `key_name as u16`? With `#[repr(u16)]` on the enum, that's safe to do, and doesn't require a language feature.

Reply:
good catch, removed use of `discriminant_value()` in favor of `as u16`